File: src/hybrid_aStar/scripts/frontend_street_map_env.py
#!/usr/bin/env python3

import math
from pathlib import Path
from typing import Tuple

import numpy as np
import rospy
from geometry_msgs.msg import Point
from nav_msgs.msg import OccupancyGrid
from planning_msgs.msg import car_info
from visualization_msgs.msg import Marker, MarkerArray


class FrontendStreetMapEnvironment:
    def __init__(self) -> None:
        self.world_frame = rospy.get_param("~frontend_street_map/world_frame", "velodyne")
        self.map_topic = rospy.get_param("~frontend_street_map/map_topic", "/guided_frontend_street_map")
        self.marker_topic = rospy.get_param(
            "~frontend_street_map/marker_topic", "/guided_frontend_street_map/markers"
        )
        self.car_pose_topic = rospy.get_param("~frontend_street_map/car_pose_topic", "/car_pos")
        self.dataset_path = Path(
            rospy.get_param(
                "~frontend_street_map/dataset_path",
                str(
                    Path(__file__).resolve().parents[1]
                    / "model_base_astar"
                    / "neural-astar"
                    / "planning-datasets"
                    / "data"
                    / "street"
                    / "mixed_064_moore_c16.npz"
                ),
            )
        )
        self.split = rospy.get_param("~frontend_street_map/split", "train")
        self.map_index = int(rospy.get_param("~frontend_street_map/map_index", 0))
        self.random_index = bool(rospy.get_param("~frontend_street_map/random_index", False))
        self.resolution = float(rospy.get_param("~frontend_street_map/resolution", 0.25))
        self.auto_refresh = bool(rospy.get_param("~frontend_street_map/auto_refresh", False))
        self.refresh_period = float(rospy.get_param("~frontend_street_map/refresh_period", 5.0))
        self.publish_default_start = bool(rospy.get_param("~frontend_street_map/publish_default_start", True))
        self.default_start_yaw = float(rospy.get_param("~frontend_street_map/default_start_yaw", 0.0))
        self.seed = int(rospy.get_param("~frontend_street_map/seed", 123))

        self.rng = np.random.default_rng(self.seed)
        self.maps = self._load_map_bank(self.dataset_path, self.split)
        self.map_pub = rospy.Publisher(self.map_topic, OccupancyGrid, queue_size=1, latch=True)
        self.marker_pub = rospy.Publisher(self.marker_topic, MarkerArray, queue_size=1, latch=True)
        self.car_pose_pub = rospy.Publisher(self.car_pose_topic, car_info, queue_size=1, latch=True)

        self.current_map_index = 0
        self.publish_map()
        if self.auto_refresh:
            rospy.Timer(rospy.Duration(self.refresh_period), self._timer_callback)

    def _load_map_bank(self, dataset_path: Path, split: str) -> np.ndarray:
        if not dataset_path.exists():
            raise FileNotFoundError(f"street dataset not found: {dataset_path}")
        data = np.load(str(dataset_path))
        split_key = {"train": "arr_0", "valid": "arr_4", "test": "arr_8"}[split]
        if split_key not in data.files:
            raise KeyError(f"{split_key} not found in {dataset_path}")
        maps = np.asarray(data[split_key], dtype=np.float32)
        if maps.ndim != 3:
            raise ValueError(f"{split_key} must be [N,H,W], got {maps.shape}")
        # planning-datasets arr_* street maps use 1=free, 0=obstacle.
        return 1.0 - maps

    def _pick_map_index(self) -> int:
        if self.random_index:
            return int(self.rng.integers(0, self.maps.shape[0]))
        return max(0, min(int(self.map_index), int(self.maps.shape[0] - 1)))

    def _origin(self, width: int, height: int) -> Tuple[float, float]:
        return -0.5 * width * self.resolution, -0.5 * height * self.resolution

    def _cell_center(self, x: int, y: int, origin_x: float, origin_y: float, z: float = 0.0) -> Point:
        p = Point()
        p.x = origin_x + (x + 0.5) * self.resolution
        p.y = origin_y + (y + 0.5) * self.resolution
        p.z = z
        return p

    def _publish_default_start(self, occ_map: np.ndarray, origin_x: float, origin_y: float) -> None:
        if not self.publish_default_start:
            return
        free_y, free_x = np.where(occ_map < 0.5)
        if free_x.size == 0:
            rospy.logwarn("street map has no free cells for default start pose")
            return
        pick = int(self.rng.integers(0, free_x.size))
        x = int(free_x[pick])
        y = int(free_y[pick])
        pt = self._cell_center(x, y, origin_x, origin_y)

        msg = car_info()
        msg.header.stamp = rospy.Time.now()
        msg.header.frame_id = self.world_frame
        msg.x = pt.x
        msg.y = pt.y
        msg.yaw = self.default_start_yaw
        msg.speedDrietion = 1
        msg.speed = 0.0
        msg.turnAngle = 0.0
        msg.yawrate = 0.0
        self.car_pose_pub.publish(msg)

    def publish_map(self) -> None:
        self.current_map_index = self._pick_map_index()
        occ_map = self.maps[self.current_map_index]
        height, width = occ_map.shape
        origin_x, origin_y = self._origin(width, height)

        grid = OccupancyGrid()
        grid.header.stamp = rospy.Time.now()
        grid.header.frame_id = self.world_frame
        grid.info.map_load_time = grid.header.stamp
        grid.info.resolution = self.resolution
        grid.info.width = width
        grid.info.height = height
        grid.info.origin.position.x = origin_x
        grid.info.origin.position.y = origin_y
        grid.info.origin.orientation.w = 1.0
        grid.data = [100 if v >= 0.5 else 0 for v in occ_map.reshape(-1)]
        self.map_pub.publish(grid)

        markers = MarkerArray()

        obstacle_marker = Marker()
        obstacle_marker.header.frame_id = self.world_frame
        obstacle_marker.header.stamp = grid.header.stamp
        obstacle_marker.ns = "frontend_street_map"
        obstacle_marker.id = 0
        obstacle_marker.type = Marker.CUBE_LIST
        obstacle_marker.action = Marker.ADD
        obstacle_marker.pose.orientation.w = 1.0
        obstacle_marker.scale.x = self.resolution
        obstacle_marker.scale.y = self.resolution
        obstacle_marker.scale.z = 0.05
        obstacle_marker.color.r = 0.62
        obstacle_marker.color.g = 0.62
        obstacle_marker.color.b = 0.62
        obstacle_marker.color.a = 0.95
        for y in range(height):
            for x in range(width):
                if occ_map[y, x] >= 0.5:
                    obstacle_marker.points.append(self._cell_center(x, y, origin_x, origin_y))
        markers.markers.append(obstacle_marker)

        boundary = Marker()
        boundary.header.frame_id = self.world_frame
        boundary.header.stamp = grid.header.stamp
        boundary.ns = "frontend_street_map"
        boundary.id = 1
        boundary.type = Marker.LINE_STRIP
        boundary.action = Marker.ADD
        boundary.pose.orientation.w = 1.0
        boundary.scale.x = 0.08
        boundary.color.r = 0.35
        boundary.color.g = 0.35
        boundary.color.b = 0.35
        boundary.color.a = 1.0
        boundary.points = [
            self._cell_center(0, 0, origin_x, origin_y),
            self._cell_center(width - 1, 0, origin_x, origin_y),
            self._cell_center(width - 1, height - 1, origin_x, origin_y),
            self._cell_center(0, height - 1, origin_x, origin_y),
            self._cell_center(0, 0, origin_x, origin_y),
        ]
        markers.markers.append(boundary)
        self.marker_pub.publish(markers)

        self._publish_default_start(occ_map, origin_x, origin_y)
        rospy.loginfo(
            "Published street map split=%s index=%d size=%dx%d resolution=%.3f",
            self.split,
            self.current_map_index,
            width,
            height,
            self.resolution,
        )

    def _timer_callback(self, _event) -> None:
        self.publish_map()


def main() -> None:
    rospy.init_node("frontend_street_map_env")
    FrontendStreetMapEnvironment()
    rospy.spin()


if __name__ == "__main__":
    main()

File: src/hybrid_aStar/app/frontend_street_map_env.cpp
#include <geometry_msgs/Point.h>
#include <nav_msgs/OccupancyGrid.h>
#include <planning_msgs/car_info.h>
#include <ros/package.h>
#include <ros/ros.h>
#include <visualization_msgs/Marker.h>
#include <visualization_msgs/MarkerArray.h>
#include <zip.h>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <fstream>
#include <random>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

struct StreetMaps {
    std::vector<float> data;
    int map_count{0};
    int height{0};
    int width{0};
};

uint16_t ReadLe16(const unsigned char* ptr) {
    return static_cast<uint16_t>(ptr[0]) | (static_cast<uint16_t>(ptr[1]) << 8);
}

uint32_t ReadLe32(const unsigned char* ptr) {
    return static_cast<uint32_t>(ptr[0]) | (static_cast<uint32_t>(ptr[1]) << 8) |
           (static_cast<uint32_t>(ptr[2]) << 16) | (static_cast<uint32_t>(ptr[3]) << 24);
}

std::string SplitArrayName(const std::string& split) {
    if (split == "train") {
        return "arr_0.npy";
    }
    if (split == "valid") {
        return "arr_4.npy";
    }
    return "arr_8.npy";
}

std::vector<char> ReadZipEntry(zip_t* archive, const std::string& entry_name) {
    zip_stat_t stat;
    if (zip_stat(archive, entry_name.c_str(), ZIP_FL_ENC_GUESS, &stat) != 0) {
        throw std::runtime_error("zip entry not found: " + entry_name);
    }
    zip_file_t* file = zip_fopen(archive, entry_name.c_str(), ZIP_FL_ENC_GUESS);
    if (file == nullptr) {
        throw std::runtime_error("failed to open zip entry: " + entry_name);
    }
    std::vector<char> buffer(static_cast<std::size_t>(stat.size));
    zip_int64_t total = 0;
    while (total < stat.size) {
        const zip_int64_t read_size = zip_fread(file, buffer.data() + total, stat.size - total);
        if (read_size < 0) {
            zip_fclose(file);
            throw std::runtime_error("failed reading zip entry: " + entry_name);
        }
        if (read_size == 0) {
            break;
        }
        total += read_size;
    }
    zip_fclose(file);
    if (total != stat.size) {
        throw std::runtime_error("short read for zip entry: " + entry_name);
    }
    return buffer;
}

StreetMaps ParseNpyFloat32Maps(const std::vector<char>& npy_bytes) {
    const unsigned char* raw = reinterpret_cast<const unsigned char*>(npy_bytes.data());
    const std::size_t size = npy_bytes.size();
    if (size < 16 || std::memcmp(raw, "\x93NUMPY", 6) != 0) {
        throw std::runtime_error("invalid npy header");
    }
    const uint8_t major = raw[6];
    std::size_t header_len = 0;
    std::size_t data_offset = 0;
    if (major == 1) {
        header_len = ReadLe16(raw + 8);
        data_offset = 10;
    } else if (major == 2) {
        header_len = ReadLe32(raw + 8);
        data_offset = 12;
    } else {
        throw std::runtime_error("unsupported npy version");
    }
    if (data_offset + header_len > size) {
        throw std::runtime_error("corrupted npy header length");
    }

    const std::string header(reinterpret_cast<const char*>(raw + data_offset), header_len);
    if (header.find("'descr': '<f4'") == std::string::npos &&
        header.find("\"descr\": \"<f4\"") == std::string::npos) {
        throw std::runtime_error("expected float32 npy tensor");
    }
    if (header.find("False") == std::string::npos) {
        throw std::runtime_error("fortran-order npy is unsupported");
    }

    const std::size_t shape_pos = header.find("shape");
    const std::size_t lparen = header.find('(', shape_pos);
    const std::size_t rparen = header.find(')', lparen);
    if (shape_pos == std::string::npos || lparen == std::string::npos || rparen == std::string::npos) {
        throw std::runtime_error("failed to parse npy shape");
    }
    std::string shape_text = header.substr(lparen + 1, rparen - lparen - 1);
    std::replace(shape_text.begin(), shape_text.end(), ',', ' ');
    std::stringstream ss(shape_text);

    StreetMaps maps;
    if (!(ss >> maps.map_count >> maps.height >> maps.width)) {
        throw std::runtime_error("expected shape [N,H,W] in npy");
    }
    const std::size_t payload_offset = data_offset + header_len;
    const std::size_t expected_values =
        static_cast<std::size_t>(maps.map_count) * static_cast<std::size_t>(maps.height) *
        static_cast<std::size_t>(maps.width);
    const std::size_t expected_bytes = expected_values * sizeof(float);
    if (payload_offset + expected_bytes > size) {
        throw std::runtime_error("npy payload truncated");
    }
    maps.data.resize(expected_values);
    std::memcpy(maps.data.data(), raw + payload_offset, expected_bytes);
    return maps;
}

StreetMaps LoadStreetMaps(const std::string& dataset_path, const std::string& split) {
    int err = 0;
    zip_t* archive = zip_open(dataset_path.c_str(), ZIP_RDONLY, &err);
    if (archive == nullptr) {
        throw std::runtime_error("failed to open npz dataset: " + dataset_path);
    }
    try {
        const std::vector<char> bytes = ReadZipEntry(archive, SplitArrayName(split));
        zip_close(archive);
        return ParseNpyFloat32Maps(bytes);
    } catch (...) {
        zip_close(archive);
        throw;
    }
}

inline std::size_t FlatIndex(int x, int y, int width) {
    return static_cast<std::size_t>(y * width + x);
}

class FrontendStreetMapEnvironment {
public:
    explicit FrontendStreetMapEnvironment(ros::NodeHandle& nh)
        : nh_(nh), rng_(static_cast<std::mt19937::result_type>(seed_)) {
        LoadParams();
        maps_ = LoadStreetMaps(dataset_path_, split_);
        map_pub_ = nh_.advertise<nav_msgs::OccupancyGrid>(map_topic_, 1, true);
        marker_pub_ = nh_.advertise<visualization_msgs::MarkerArray>(marker_topic_, 1, true);
        car_pose_pub_ = nh_.advertise<planning_msgs::car_info>(car_pose_topic_, 1, true);

        PublishMap();
        if (auto_refresh_) {
            timer_ = nh_.createTimer(ros::Duration(refresh_period_),
                                     &FrontendStreetMapEnvironment::TimerCallback,
                                     this);
        }
    }

private:
    void LoadParams() {
        world_frame_ = nh_.param<std::string>("frontend_street_map/world_frame", "velodyne");
        map_topic_ = nh_.param<std::string>("frontend_street_map/map_topic", "/guided_frontend_street_map");
        marker_topic_ =
            nh_.param<std::string>("frontend_street_map/marker_topic", "/guided_frontend_street_map/markers");
        car_pose_topic_ = nh_.param<std::string>("frontend_street_map/car_pose_topic", "/car_pos");
        const std::string default_dataset =
            ros::package::getPath("hybrid_a_star") +
            "/model_base_astar/neural-astar/planning-datasets/data/street/mixed_064_moore_c16.npz";
        dataset_path_ = nh_.param<std::string>("frontend_street_map/dataset_path", default_dataset);
        split_ = nh_.param<std::string>("frontend_street_map/split", "train");
        map_index_ = nh_.param("frontend_street_map/map_index", 0);
        random_index_ = nh_.param("frontend_street_map/random_index", false);
        resolution_ = nh_.param("frontend_street_map/resolution", 0.25);
        auto_refresh_ = nh_.param("frontend_street_map/auto_refresh", false);
        refresh_period_ = nh_.param("frontend_street_map/refresh_period", 5.0);
        publish_default_start_ = nh_.param("frontend_street_map/publish_default_start", true);
        default_start_yaw_ = nh_.param("frontend_street_map/default_start_yaw", 0.0);
        seed_ = nh_.param("frontend_street_map/seed", 123);
        rng_.seed(static_cast<std::mt19937::result_type>(seed_));
    }

    int PickMapIndex() {
        if (random_index_) {
            std::uniform_int_distribution<int> dis(0, maps_.map_count - 1);
            return dis(rng_);
        }
        return std::max(0, std::min(map_index_, maps_.map_count - 1));
    }

    geometry_msgs::Point CellCenter(int x, int y, double z = 0.0) const {
        geometry_msgs::Point p;
        p.x = origin_x_ + (static_cast<double>(x) + 0.5) * resolution_;
        p.y = origin_y_ + (static_cast<double>(y) + 0.5) * resolution_;
        p.z = z;
        return p;
    }

    void PublishDefaultStart() {
        if (!publish_default_start_) {
            return;
        }

        std::vector<std::pair<int, int>> free_cells;
        free_cells.reserve(static_cast<std::size_t>(width_ * height_));
        for (int y = 0; y < height_; ++y) {
            for (int x = 0; x < width_; ++x) {
                if (current_occ_[FlatIndex(x, y, width_)] == 0) {
                    free_cells.emplace_back(x, y);
                }
            }
        }
        if (free_cells.empty()) {
            ROS_WARN("street map has no free cells for default start pose");
            return;
        }

        std::uniform_int_distribution<int> dis(0, static_cast<int>(free_cells.size()) - 1);
        const auto& xy = free_cells[dis(rng_)];
        const geometry_msgs::Point pt = CellCenter(xy.first, xy.second);

        planning_msgs::car_info msg;
        msg.header.stamp = ros::Time::now();
        msg.header.frame_id = world_frame_;
        msg.x = pt.x;
        msg.y = pt.y;
        msg.yaw = default_start_yaw_;
        msg.speedDrietion = 1;
        msg.speed = 0.0;
        msg.turnAngle = 0.0;
        msg.yawrate = 0.0;
        car_pose_pub_.publish(msg);
    }

    void PublishMap() {
        current_map_index_ = PickMapIndex();
        width_ = maps_.width;
        height_ = maps_.height;
        origin_x_ = -0.5 * static_cast<double>(width_) * resolution_;
        origin_y_ = -0.5 * static_cast<double>(height_) * resolution_;
        current_occ_.assign(static_cast<std::size_t>(width_ * height_), 0);

        const std::size_t offset =
            static_cast<std::size_t>(current_map_index_) * static_cast<std::size_t>(width_ * height_);
        for (int y = 0; y < height_; ++y) {
            for (int x = 0; x < width_; ++x) {
                const float free_value = maps_.data[offset + FlatIndex(x, y, width_)];
                current_occ_[FlatIndex(x, y, width_)] = free_value > 0.5f ? 0 : 1;
            }
        }

        nav_msgs::OccupancyGrid grid;
        grid.header.stamp = ros::Time::now();
        grid.header.frame_id = world_frame_;
        grid.info.map_load_time = grid.header.stamp;
        grid.info.resolution = resolution_;
        grid.info.width = static_cast<uint32_t>(width_);
        grid.info.height = static_cast<uint32_t>(height_);
        grid.info.origin.position.x = origin_x_;
        grid.info.origin.position.y = origin_y_;
        grid.info.origin.orientation.w = 1.0;
        grid.data.resize(current_occ_.size());
        for (std::size_t i = 0; i < current_occ_.size(); ++i) {
            grid.data[i] = current_occ_[i] == 0 ? 0 : 100;
        }
        map_pub_.publish(grid);

        visualization_msgs::MarkerArray markers;

        visualization_msgs::Marker obstacle_marker;
        obstacle_marker.header.frame_id = world_frame_;
        obstacle_marker.header.stamp = grid.header.stamp;
        obstacle_marker.ns = "frontend_street_map";
        obstacle_marker.id = 0;
        obstacle_marker.type = visualization_msgs::Marker::CUBE_LIST;
        obstacle_marker.action = visualization_msgs::Marker::ADD;
        obstacle_marker.pose.orientation.w = 1.0;
        obstacle_marker.scale.x = resolution_;
        obstacle_marker.scale.y = resolution_;
        obstacle_marker.scale.z = 0.05;
        obstacle_marker.color.r = 0.62;
        obstacle_marker.color.g = 0.62;
        obstacle_marker.color.b = 0.62;
        obstacle_marker.color.a = 0.95;
        for (int y = 0; y < height_; ++y) {
            for (int x = 0; x < width_; ++x) {
                if (current_occ_[FlatIndex(x, y, width_)] != 0) {
                    obstacle_marker.points.push_back(CellCenter(x, y));
                }
            }
        }
        markers.markers.push_back(obstacle_marker);

        visualization_msgs::Marker boundary;
        boundary.header.frame_id = world_frame_;
        boundary.header.stamp = grid.header.stamp;
        boundary.ns = "frontend_street_map";
        boundary.id = 1;
        boundary.type = visualization_msgs::Marker::LINE_STRIP;
        boundary.action = visualization_msgs::Marker::ADD;
        boundary.pose.orientation.w = 1.0;
        boundary.scale.x = 0.08;
        boundary.color.r = 0.35;
        boundary.color.g = 0.35;
        boundary.color.b = 0.35;
        boundary.color.a = 1.0;
        boundary.points = {
            CellCenter(0, 0),
            CellCenter(width_ - 1, 0),
            CellCenter(width_ - 1, height_ - 1),
            CellCenter(0, height_ - 1),
            CellCenter(0, 0),
        };
        markers.markers.push_back(boundary);
        marker_pub_.publish(markers);

        PublishDefaultStart();
        ROS_INFO("Published C++ street map split=%s index=%d size=%dx%d resolution=%.3f",
                 split_.c_str(),
                 current_map_index_,
                 width_,
                 height_,
                 resolution_);
    }

    void TimerCallback(const ros::TimerEvent&) {
        PublishMap();
    }

    ros::NodeHandle nh_;
    ros::Publisher map_pub_;
    ros::Publisher marker_pub_;
    ros::Publisher car_pose_pub_;
    ros::Timer timer_;

    std::string world_frame_{"velodyne"};
    std::string map_topic_{"/guided_frontend_street_map"};
    std::string marker_topic_{"/guided_frontend_street_map/markers"};
    std::string car_pose_topic_{"/car_pos"};
    std::string dataset_path_;
    std::string split_{"train"};
    int map_index_{0};
    bool random_index_{false};
    double resolution_{0.25};
    bool auto_refresh_{false};
    double refresh_period_{5.0};
    bool publish_default_start_{true};
    double default_start_yaw_{0.0};
    int seed_{123};

    StreetMaps maps_;
    std::mt19937 rng_;
    int current_map_index_{0};
    int width_{0};
    int height_{0};
    double origin_x_{0.0};
    double origin_y_{0.0};
    std::vector<int> current_occ_;
};

}  // namespace

int main(int argc, char** argv) {
    ros::init(argc, argv, "frontend_street_map_env_cpp");
    ros::NodeHandle nh("~");
    try {
        FrontendStreetMapEnvironment env(nh);
        ros::spin();
    } catch (const std::exception& e) {
        ROS_FATAL_STREAM("frontend_street_map_env_cpp failed: " << e.what());
        return 1;
    }
    return 0;
}
